packages/flutter_tools/bin/xcode_backend.dart

```diff
@@ -326,6 +350,92 @@ class Context {
+  bool _validateFlutterFramework({
+    required String buildMode,
+    required TargetPlatform platform,
+    required String? builtProductsDir,
+    required String? targetBuildDir,
+  }) {
+    if (builtProductsDir == null || targetBuildDir == null) {
+      return false;
+    }
+    try {
+      final infoPlistFromBuild = File('$builtProductsDir/${platform.infoPlistPath}');
+      if (!infoPlistFromBuild.existsSync()) {
+        return false;
+      }
+      final infoPlistFromEmbedded = File('$targetBuildDir/${platform.infoPlistPath}');
+      if (!infoPlistFromEmbedded.existsSync()) {
+        return false;
+      }
+
+      final File? infoPlistFromEngineCache = _infoPlistFromEngineCache(buildMode, platform);
+      if (infoPlistFromEngineCache == null || !infoPlistFromEngineCache.existsSync()) {
+        return false;
+      }
+      final String expectedInfoPlist = infoPlistFromEngineCache.readAsStringSync();
+      return infoPlistFromBuild.readAsStringSync() == expectedInfoPlist &&
+          infoPlistFromEmbedded.readAsStringSync() == expectedInfoPlist;
+    } on Exception catch (e) {
+      // Use `echo` instead of `echoError` so it does not cause the build to fail.
+      echo('An error occured while validating the Flutter framework: $e');
+    }
+    return false;
+  }
+
+  /// Find the Info.plist of the Flutter/FlutterMacOS framework for the corresponding [buildMode]
+  /// and [platform].
+  File? _infoPlistFromEngineCache(String buildMode, TargetPlatform platform) {
+    final String artifactMode = buildMode == 'debug'
+        ? platform.artifactName
+        : '${platform.artifactName}-$buildMode';
+    final xcframework = Directory(
+      '${environment['FLUTTER_ROOT'] ?? ''}/bin/cache/artifacts/engine/$artifactMode/${platform.frameworkName}.xcframework',
+    );
+    switch (platform) {
+      case TargetPlatform.ios:
+        final String? sdkRoot = environment['SDKROOT']?.toLowerCase();
+        if (sdkRoot == null || !sdkRoot.contains('iphone')) {
+          return null;
+        }
+        final bool simulatorSDK = sdkRoot.contains('simulator');
+        for (final FileSystemEntity entity in xcframework.listSync()) {
+          final String platformBaseName = Uri.parse(entity.path).pathSegments.last;
+          if (entity is Directory && platformBaseName.startsWith('ios-')) {
+            final bool isSimulatorDirectory = platformBaseName.endsWith('-simulator');
+            if (simulatorSDK == isSimulatorDirectory) {
+              return File('${entity.path}/${platform.infoPlistPath}');
+            }
+          }
+        }
+      case TargetPlatform.macos:
+        for (final FileSystemEntity entity in xcframework.listSync()) {
+          final String platformBaseName = Uri.parse(entity.path).pathSegments.last;
+          if (entity is Directory && platformBaseName.startsWith('macos-')) {
+            return File('${entity.path}/${platform.infoPlistPath}');
+          }
+        }
```

Comment:
This is mimicking:

https://github.com/flutter/flutter/blob/f63cf73d310065a11b67d82ba88618e342227c7c/packages/flutter_tools/lib/src/artifacts.dart#L1026-L1033